### PR TITLE
Add cluster labels

### LINKED ISSUE #3604: Provide option to set `cluster_display_name`
### Feature Request

We currently recommend that people set cluster name to something meaningful, and since we show this in the UI and CLI we should provide an easy way to change the name without having to wipe out `/var/lib/teleport`.

```yaml
# This section configures the 'auth service':
auth_service:
    # Turns 'auth' role on. Default is 'yes'
    enabled: yes

    # A cluster name is used as part of a signature in certificates
    # generated by this CA.
    #
    # We strongly recommend to explicitly set it to something meaningful as it
    # becomes important when configuring trust between multiple clusters.
    #
    # By default an automatically generated name is used (not recommended)
    #
    # IMPORTANT: if you change cluster_name, it will invalidate all generated
    # certificates and keys (may need to wipe out /var/lib/teleport directory)
    cluster_name: "main"
    cluster_display: "Acme Cluster"
    # See explanation of labels in "Labeling Clusters" section below
    labels:
        location: azure
        type: MSP 
        status: standby
```

https://github.com/gravitational/teleport/issues/2317

### Motivation
I was reminded about this from https://github.com/gravitational/teleport/issues/2317, as we show more UI for trusted cluster and expand the CLI these extra labels and context will improve the Ops/Dev Experience. 

### Who's it for?
OSS User, Pro, Enterprise

### PR DESCRIPTION
**Pull request by klizhentas**
Fixes #3604

This commit adds support for cluster_labels
role parameter limiting access to remote clusters by label.
New tctl update rc provides interface to set labels on remote clusters.

Consider two clusers, `one` - root and `remote` - leaf.

```bash
$ tsh clusters
Cluster Name Status
------------ ------
one          online
two          online
```

Create the trusted cluster join token with labels:

```bash
$ tctl tokens add --type=trusted_cluster --labels=env=prod
```

Every cluster joined using this token will inherit env:prod labels.

Alternatively, update remote cluster labels by modifying
`rc` command. Letting remote clusters to propagate their labels
creates a problem of rogue clusters updating their labels to bad values.

Instead, administrator of root cluster control the labels
using remote clusters API without fear of override:

```bash
$ tctl get rc

kind: remote_cluster
metadata:
  name: two
status:
  connection: online
  last_heartbeat: "2020-09-14T03:13:59.35518164Z"
version: v3
```

```bash
$ tctl update rc/two --set-labels=env=prod

cluster two has been updated
```

```bash
$ tctl get rc
kind: remote_cluster
metadata:
  labels:
    env: prod
  name: two
status:
  connection: online
  last_heartbeat: "2020-09-14T03:13:59.35518164Z"
```

Update the role to deny access to prod env:

```yaml
kind: role
metadata:
  name: dev
spec:
  allow:
    logins: [root]
    node_labels:
      '*': '*'

    # Cluster labels control what clusters user can connect to. The wildcard ('*') means
    # any cluster. If no role in the role set is using labels and cluster is not labeled,
    # the cluster labels check is not applied. Otherwise, cluster labels are always enforced.
    # This makes the feature backwards-compatible.
    cluster_labels:
      'env': 'staging'
  deny:
    # cluster labels control what clusters user can connect to. The wildcard ('*') means
    # any cluster. By default none is set in deny rules to preserve backwards compatibility
    cluster_labels:
      'env': 'prod'
```

```bash
$ tctl create -f dev.yaml
```

Cluster two is now invisible to user with `dev` role.

```bash
$ tsh clusters
Cluster Name Status
------------ ------
one          online
```

### CODE REVIEW

**Comment by klizhentas** [comment]
@fspmarshall @russjones @awly I will add more unit tests, but integration tests and feature is ready for your review. Please folks take a look as we have to deliver this feature in beta 5.0 by the end of the week.

File: lib/services/types.proto
Comment (fspmarshall) [inline lib/services/types.proto:760]: I wonder of we should call this something like `remote_cluster_labels` or `leaf_cluster_labels` to better clarify that these labels don't affect the user's access to the cluster in which the role is defined.  Its obvious that this is true if you understand teleport's internals, but may not be obvious to all users.

**Review by fspmarshall — COMMENTED** [review]
Are we deliberately not supporting cluster label interpolation in roles to start with?  If not, I think it'd be best to standardize the labels s.t. node and cluster labels behave identically wherever possible.

File: lib/services/types.proto
Comment (klizhentas) [reply lib/services/types.proto:760]: I was thinking to extend the cluster labels feature to main cluster as well, we just lack controls over the cluster labels at the moment, but once implemented will complete the cycle :)

File: lib/cache/cache_test.go
Comment (awly) [inline lib/cache/cache_test.go:1196]: optional: please use the new testing style

File: lib/cache/cache_test.go
Comment (awly) [inline lib/cache/cache_test.go:1236]: nit: check that labels are set here

File: lib/reversetunnel/api_with_roles.go
Comment (awly) [inline lib/reversetunnel/api_with_roles.go:42]:
```suggestion
```

File: lib/reversetunnel/api_with_roles.go
Comment (awly) [inline lib/reversetunnel/api_with_roles.go:58]: this is O(N) requests to the auth server
can you fetch all remote clusters and filter them against local `RemoteSite`s?
or will this always be served from the cache?

File: lib/services/local/presence.go
Comment (awly) [inline lib/services/local/presence.go:614]: why not update all the fields?
this seems like maintenance burden

File: lib/services/oidc_test.go
Comment (awly) [inline lib/services/oidc_test.go:61]: optional: do a deep compare on the entire `oc` struct

File: lib/services/oidc_test.go
Comment (awly) [inline lib/services/oidc_test.go:28]: optional: merge all these unmarshal tests into a single table-driven test

File: lib/services/role.go
Comment (awly) [inline lib/services/role.go:1530]:
```suggestion
	// is not labelled, assume that the role set has access to the cluster.
```

File: lib/services/role.go
Comment (awly) [inline lib/services/role.go:1556]:
```suggestion
		if matchLabels {
```

File: lib/services/role.go
Comment (awly) [inline lib/services/role.go:1573]: optional: move all these values to `log.WithFields`

File: lib/services/role.go
Comment (awly) [inline lib/services/role.go:1557]: this check shouldn't be necessary, since you call `log.Debugf`

File: lib/services/role.go
Comment (awly) [inline lib/services/role.go:1578]:
```suggestion
		if matchLabels {
```

File: lib/services/role.go
Comment (awly) [inline lib/services/role.go:1588]: same here, remove this check

File: lib/services/types.proto
Comment (awly) [inline lib/services/types.proto:757]: shouldn't this field be nullable?

File: lib/web/apiserver.go
Comment (awly) [inline lib/web/apiserver.go:1361]:
```suggestion
	clusters := append(remoteClusters, rc)
```

File: lib/web/ui/cluster.go
Comment (awly) [inline lib/web/ui/cluster.go:74]:
```suggestion
	clusters := make([]Cluster, 0, len(remoteClusters))
```

File: tool/tctl/common/resource_command.go
Comment (awly) [inline tool/tctl/common/resource_command.go:444]:
```suggestion
		return trace.BadParameter("provide a full resource name to update, for example:\n$ tctl update rc/remote\n")
```

File: tool/tctl/common/resource_command.go
Comment (awly) [inline tool/tctl/common/resource_command.go:456]: you can use a non-pointer `time.Time` and call `expiry.IsZero()` to see if it's set

**Review by awly — COMMENTED** [review]
What is the risk of letting remote clusters define their own labels? An example attack/escalation would be helpful to understand.
If there is no concrete risk, I'd recommend keeping this consistent with nodes and letting remote clusters define their own labels.

Also, please update the PR description to mention that `role_template` is removed from the config, in case something breaks and we need to find the culprit PR.

**Comment by klizhentas** [comment]
> What is the risk of letting remote clusters define their own labels? An example attack/escalation would be helpful to understand.

If there is no concrete risk, I'd recommend keeping this consistent with nodes and letting remote clusters define their own labels.

Customers are often used in multi-tenant environments. In this case, leaf clusters can't be allowed to set their own labels, because they will impersonate another tenant.

Company that needs this feature operates in multi-tenant environment, that's why I'm building this feature in this way right from the start. I would also like to add similar feature for nodes as well.  

Second problem is that upgrading remote clusters should be easy and automated from the root clusters. In alternative scenario, using static node-like configuration is that administrators will have to upgrade every single auth server and restart it.

File: lib/auth/auth.go
Comment (klizhentas) [reply lib/auth/auth.go:849]: Yes, exactly. Nodes and proxies should inherit labels from tokens encoded in the X509 metada that they can't change. This will allow scenarios when nodes are deployed to other tenants. It is less relevant for proxies.

File: lib/services/role.go
Comment (klizhentas) [reply lib/services/role.go:1600]: @awly so we are no longer doing `matchLabels == true`?

File: lib/services/local/presence.go
Comment (klizhentas) [reply lib/services/local/presence.go:614]: to avoid updating parts that should not be updated - like connection status or last heartbeat

File: lib/reversetunnel/api_with_roles.go
Comment (klizhentas) [reply lib/reversetunnel/api_with_roles.go:58]: yes, this is always served from cache that is using in memory `BTree`

**Comment by klizhentas** [comment]
@fspmarshall I've fixed the interpolation for cluster labels, good catch

File: lib/services/role.go
Comment (klizhentas) [reply lib/services/role.go:1579]: this is to avoid formatting call on a large scale

File: lib/services/local/presence.go
Comment (awly) [reply lib/services/local/presence.go:614]: How about overwriting those fields with the current backend values then?
Instead of copying over just the fields that can be updated, suppress the changes to the fields that can't be updated.

File: lib/services/role.go
Comment (awly) [reply lib/services/role.go:1600]: Nope, it was never preferred.
The only places where we keep that is for truth tables (exhaustive switches with multiple bools), because a few folks on the team prefer that.

File: lib/services/role.go
Comment (awly) [reply lib/services/role.go:1579]: https://github.com/sirupsen/logrus/blob/d131c24e23baaa812461202af6d7cfa388e2d292/entry.go#L328-L330 looks like the formatting call doesn't happen when log level check fails

**Comment by knisbet** [comment]
I'm not looking at the whole PR, just to add some food for thought about trust and labels with the called out comment. 

I believe the approach kubernetes is taking for the untrusted labels problem, is to slowly create a distinction between well known labels that are security relevant that a node cannot control, and more optional labels or extensions a node can control, such as to help facilitate something like service discovery. Although kubernetes is starting from the point where it was insecure, and are now trying to add this in later which is challenging. Another option is to namespace insecure labels or to call them something different like annotations or metadata instead of labels, that have an implicit security difference, although these sort of clever solutions do create the possibility for a user to get confused or forgot about the security implications.

File: tool/tctl/common/resource_command.go
Comment (r0mant) [inline tool/tctl/common/resource_command.go:480]: Nit: This block (get metadata, set labels, set metadata) has appeared at least 3 times in this PR making me think it might be worth providing `SetLabels()` shorthand on the resource interface.

File: tool/tctl/common/resource_command.go
Comment (r0mant) [inline tool/tctl/common/resource_command.go:490]: Nit: Include resource types that support updates in the message?

File: tool/tctl/common/resource_command.go
Comment (r0mant) [inline tool/tctl/common/resource_command.go:470]: Nit: Since this is a top-level context for the command, I think using `context.Background()` instead of `TODO()` may be justified in this case.

File: lib/services/types.proto
Comment (klizhentas) [reply lib/services/types.proto:757]: because we are using custom type, I think it will generate *LabelValues

**Comment by klizhentas** [comment]
@fspmarshall @r0mant @awly great review, please take another look

File: lib/client/client.go
Comment (awly) [inline lib/client/client.go:137]:
```suggestion
	// Local cluster should always be first in the list.
	out := []services.RemoteCluster{rc}
```

File: lib/client/client.go
Comment (awly) [inline lib/client/client.go:212]: if I remember correctly, a client can run with a nil agent (e.g. when using an identity file)
it has caused a few bugs in the past, with nil pointer dereferences.

File: lib/reversetunnel/api_with_roles.go
Comment (awly) [inline lib/reversetunnel/api_with_roles.go:63]:
```suggestion
```

File: lib/reversetunnel/api_with_roles.go
Comment (awly) [inline lib/reversetunnel/api_with_roles.go:56]: inject local cluster as first element, to be consistent

File: lib/reversetunnel/api_with_roles.go
Comment (awly) [inline lib/reversetunnel/api_with_roles.go:88]:
```suggestion
```

File: lib/services/local/presence.go
Comment (awly) [inline lib/services/local/presence.go:627]:
```suggestion
	update.SetMetadata(rc.GetMetadata())
```

File: lib/services/role.go
Comment (awly) [inline lib/services/role.go:1545]:
```suggestion
// adding logging to this function (which is called on every cluster returned
// by GetRemoteClusters) can slow down this function by 50x for large clusters!
```

File: lib/services/role.go
Comment (awly) [reply lib/services/role.go:1584]: I still think this check is redundant (throughout this method)

File: lib/web/apiserver.go
Comment (awly) [inline lib/web/apiserver.go:1360]:
```suggestion
	// Local cluster should always come first in the list.
	clusters := []services.RemoteCluster{rc}
```

File: lib/web/ui/cluster.go
Comment (awly) [inline lib/web/ui/cluster.go:74]:
```suggestion
	clusters := make([]Cluster, 0, len(remoteClusters))
```

File: lib/web/ui/cluster.go
Comment (awly) [inline lib/web/ui/cluster.go:86]: nit: this does not keep the root cluster first in the list

File: tool/tctl/common/resource_command.go
Comment (awly) [reply tool/tctl/common/resource_command.go:456]: ping

**Review by fspmarshall — APPROVED** [review]

File: lib/cache/cache.go
Comment (russjones) [inline lib/cache/cache.go:72]: Have you tested this with older clusters? I ran into a problem when I was adding support for `services.AppServer` in Application Access. See now I fixed it in this commit:

https://github.com/gravitational/teleport/pull/4582/commits/0700990077d59fce75f91bcad140c8de1e438851

File: tool/tctl/common/resource_command.go
Comment (russjones) [inline tool/tctl/common/resource_command.go:102]: We don't sorry anything like this at the moment, using command line flags to update the resource. Is this a change we want to make? Add convenience flags?

File: lib/services/role.go
Comment (a-palchikov) [inline lib/services/role.go:1565]: `!usesLabels` otherwise this might be flagged by the linter.

File: lib/services/types.proto
Comment (a-palchikov) [inline lib/services/types.proto:1098]: nit: Version is resource version

File: lib/client/client.go
Comment (klizhentas) [reply lib/client/client.go:212]: it's also used in `ConnectToCluster` and `ReissueCerts` and I'm not quite sure what to use instead if this is not available. I think you are right that `tsh --i clusters ls` will fail with or without this addition. I have created a separate ticket to fix this and other instances. https://github.com/gravitational/teleport/issues/4669

File: lib/reversetunnel/api_with_roles.go
Comment (klizhentas) [reply lib/reversetunnel/api_with_roles.go:56]: it's always first in this package as returned by (s *server) in srv.go

File: lib/services/local/presence.go
Comment (klizhentas) [reply lib/services/local/presence.go:627]: I don't want to update everything, just labels

File: lib/cache/cache.go
Comment (klizhentas) [reply lib/cache/cache.go:72]: I will merge my PR after you merge application access to get a magic fix :)

File: tool/tctl/common/resource_command.go
Comment (klizhentas) [reply tool/tctl/common/resource_command.go:102]: yes, why not? `--set-labels` and `--set-ttl` is common enough to update labels or TTL on resources

File: lib/services/role.go
Comment (klizhentas) [reply lib/services/role.go:1565]: we have the linter exception for this clause on

File: lib/web/ui/cluster.go
Comment (alex-kovoy) [reply lib/web/ui/cluster.go:86]: We do client-side sorting so this is not really needed.

File: lib/services/local/presence.go
Comment (awly) [reply lib/services/local/presence.go:627]: Ahh, in that case:
```suggestion
	meta := update.GetMetadata()
	meta.Labels = rc.GetMetadata().Labels
	update.SetMetadata(meta)
```

File: lib/services/role.go
Comment (awly) [reply lib/services/role.go:1584]: https://github.com/gravitational/teleport/issues/4681

**Review by awly — APPROVED** [review]

**Review by russjones — APPROVED** [review]

**Pull request merged**
